sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -171,6 +171,63 @@ case class InSubqueryExec(
   }
 }
 
+/**
+ * The physical node of exists-subquery. This is for support use exists in join's on condition,
+ * since some join type we can't pushdown exists condition, we plan it here
+ */
+case class ExistsExec(child: Expression,
+                      subQuery: String,
+                      plan: BaseSubqueryExec,
+                      exprId: ExprId,
+                      private var resultBroadcast: Broadcast[Boolean] = null)
+  extends ExecSubqueryExpression {
+
+  @transient private var result: Boolean = _
+
+  override def dataType: DataType = BooleanType
+  override def children: Seq[Expression] = child :: Nil
+  override def nullable: Boolean = child.nullable
+  override def toString: String = s"EXISTS ${plan.name}"
+  override def withNewPlan(plan: BaseSubqueryExec): ExistsExec = copy(plan = plan)
+
+  override def semanticEquals(other: Expression): Boolean = other match {
+    case in: ExistsExec => child.semanticEquals(in.child) && plan.sameResult(in.plan)
+    case _ => false
+  }
+
+
+  def updateResult(): Unit = {
+    result = !plan.execute().isEmpty()
+    resultBroadcast = plan.sqlContext.sparkContext.broadcast[Boolean](result)
+  }
+
+  def values(): Option[Boolean] = Option(resultBroadcast).map(_.value)
+
+  private def prepareResult(): Unit = {
+    require(resultBroadcast != null, s"$this has not finished")
+    result = resultBroadcast.value
+  }
+
+  override def eval(input: InternalRow): Any = {
+    prepareResult()
+    result
+  }
+
+  override lazy val canonicalized: ExistsExec = {
+    copy(
+      child = child.canonicalized,
+      subQuery = subQuery,
+      plan = plan.canonicalized.asInstanceOf[BaseSubqueryExec],
+      exprId = ExprId(0),
+      resultBroadcast = null)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    prepareResult()
+    ExistsSubquery(child, subQuery, result).doGenCode(ctx, ev)
```

Comment:
why we create `ExistsSubquery` to only do codegen? can we put the codegen logic in `ExistsExec`?

Reply:
> why we create `ExistsSubquery` to only do codegen? can we put the codegen logic in `ExistsExec`?

There are conflicts between ExecSubqueryExpression and UnaryExpression， 
They are both abstract class.

Reply:
We don't have to extend `UnaryExpression` and we can still implement codegen, right?

Reply:
> We don't have to extend `UnaryExpression` and we can still implement codegen, right?

Done